rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -82,6 +88,38 @@ struct AnySubscriptionCallback
   {
     shared_ptr_with_info_callback = callback;
   }
+
+  template<typename CallbackT,
+  typename std::enable_if<
+    function_traits<CallbackT>::arity == 1
+  >::type * = nullptr,
+  typename std::enable_if<
+    std::is_same<
+      typename function_traits<CallbackT>::template argument_type<0>,
+      typename std::shared_ptr<const MessageT>
+    >::value
+  >::type * = nullptr
+  >
+  void set(CallbackT callback)
+  {
+    const_shared_ptr_callback = callback;
+  }
+
+  template<typename CallbackT,
+  typename std::enable_if<
+    function_traits<CallbackT>::arity == 2
+  >::type * = nullptr,
+  typename std::enable_if<
+    std::is_same<
+      typename function_traits<CallbackT>::template argument_type<0>,
+      typename std::shared_ptr<const MessageT>
+    >::value
+  >::type * = nullptr
+  >
+  void set(CallbackT callback)
+  {
```

Comment:
Still confused about this. I think we do care about the type as well as the arity of the callback here. If Windows were only looking at the arity and not the type, then I would expect any code that exercises the new callback type to fail, because the subscriber would end up trying to call the `shared_ptr` callback instead of the `const_shared_ptr` callback...

In fact, I'm added a test for this since the new callback type to make sure:

ros2/system_tests#53

http://ci.ros2.org/job/ros2_batch_ci_windows/460/